### PR TITLE
feat: review of the supported cryptographic keys

### PR DESCRIPTION
**Pull request by tranchitella**
Changelog: None
Ticket: MEN-7535

### CODE REVIEW

File: 02.Overview/11.Security/docs.md [hunk 1]
```diff
@@ -34,14 +34,14 @@ the Mender Client can connect to the server over HTTPS, you can schedule updates
...
 
-Each Device has a unique public and private RSA, ECDSA or ED25519 key. You can
-generate this offline and provision it with the Device storage, otherwise the
-Mender Client will automatically generate an RSA key pair with a default length
-of 3072 bits when it launches for the first time. Once generated, private key cannot be
-changed or retrieved by means of API calls. If you decide to re-generate the
-keys on the Device, it will require going through the authorization process
-again. The client passes the public key in authorization requests to the server and
-you can see the public key of a Device in the Mender UI.
+Each Device has a unique public and private RSA, ECDSA with curve P-256 or ED25519
```
Comment (estenberg) [inline 02.Overview/11.Security/docs.md:37]: Strictly speaking, ED25519 is not ECDSA, and ECDSA is a signature algorithm (not a key type). So this was wrong to start off with.

Each Device has a unique RSA or Elliptic Curve (P-256 or Ed25519) keypair (public and private key). 

Strictly-strictly speaking it is a Curve25519 keypair, but I see it's often referred to as Ed25519 keypair, so we should be fine with that.

File: 02.Overview/11.Security/docs.md [hunk 2]
```diff
@@ -34,14 +34,14 @@ the Mender Client can connect to the server over HTTPS, you can schedule updates
...
+key. You can generate this offline and provision it with the Device storage, otherwise
+the Mender Client will automatically generate an Ed25519 key pair when it launches for
+the first time. Once generated, private key cannot be changed or retrieved by means
+of API calls. If you decide to re-generate the keys on the Device, it will require
+going through the authorization process again. The client passes the public key in
+authorization requests to the server and you can see the public key of a Device
+in the Mender UI.
```
Comment (estenberg) [inline 02.Overview/11.Security/docs.md:45]: Suggesting to add a note here:
(with right amount of exlamation marks for versioning / introuction compatibility note, don't quite recall :)

!! Mender Client version 4.1.0 and later generates an Ed25519 keypair by default, while prior versions generated a 3072 bit RSA key by default. These have the same security level (128 bit), but the Ed25519 keys are smaller and more efficient.

File: 06.Artifact-creation/07.Sign-and-verify/docs.md [hunk 3]
```diff
@@ -31,8 +31,12 @@ Multiple keys can be used with the `ArtifactVerifyKeys` option, in which case th
 ## Supported signing algorithms
 
 Mender supports the following signing algorithms:
-* RSA with recommended key length of at least 3072 bits
+
+* PKCS#1 v1.5 (RSA), more formally called `RSASSA-PKCS1-v1_5`, with recommended RSA key length of at least 3072 bits
 * ECDSA with curve P-256
+* Ed25519, only when used in combination with KeyFactor
```
Comment (estenberg) [inline 06.Artifact-creation/07.Sign-and-verify/docs.md:37]: Don't think this relates to KeyFactor?

Or I am not aware of that.. why can't we sign Artifacts using Ed25519 without KeyFactor?

**Review by estenberg — CHANGES_REQUESTED** [review]
Small tweaks.

File: 06.Artifact-creation/07.Sign-and-verify/docs.md (hunk 3 above)
Comment (vpodzime) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md:37]: Are you talking about Mender Artifact sources? I have no idea what `KeyFactor` is, but I see no mention of it in the Client 4 sources (or the MCU client's sources, FWIW)

Comment (estenberg) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md:37]: Ah, okey.

Then for this task I would remove that line (* Ed25519, only when used in combination with KeyFactor) as this is a bit different (it doesn't use a Ed25519 key, but a partner REST API).

Rather, add KeyFactor here: https://docs.mender.io/artifact-creation/sign-and-verify#cloud-key-management

Sorry to scope-creep, but this section should be renamed from "Supported cryptography standards" to "Hardware security" https://docs.mender.io/artifact-creation/sign-and-verify#cloud-key-management ?

Comment (lluiscampos) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md:37]: There actually might be something to this. As for the client,

We have unit tests with Ed25519 keys only for _signing_ things, which means signatures used for auth requests. I quickly wrote a unit test for _verifying_ a signature, and it fails in the OpenSSL initialization (iow when reading a Ed25519 public key).

Then we have acceptance tests for the verification of signed Artifacts, using RSA and ECDSA P-256 keys. This matches Fabio's findings, as we don't test KeyFactor integration with the actual client.

However... How can KeyFactor integration work if the client does not accept the Ed25519 public key? This sounds to me like a bug in the client (a possible regression from 3.x series).

I am creating a task for it.

File: 06.Artifact-creation/07.Sign-and-verify/docs.md [hunk 4]
```diff
@@ -31,9 +31,12 @@ Multiple keys can be used with the `ArtifactVerifyKeys` option, in which case th
 ## Supported signing algorithms
 
 Mender supports the following signing algorithms:
-* RSA with recommended key length of at least 3072 bits
+
+* PKCS#1 v1.5 (RSA), more formally called `RSASSA-PKCS1-v1_5`, with recommended RSA key length of at least 3072 bits
 * ECDSA with curve P-256
 
+On Linux, both `mender-artifact` and the Mender Client support the PKCS#11 programming interface to manipulate cryptographic tokens using `openssl`.
```
Comment (vpodzime) [inline 06.Artifact-creation/07.Sign-and-verify/docs.md:38]: Is the mention of `openssl` important here? PKCS#11 is an interface, it's not important by which means we use it, right?

File: 06.Artifact-creation/07.Sign-and-verify/docs.md [hunk 5]
```diff
@@ -117,12 +120,21 @@ For OS specific instructions on how to install and enable verification keys, vis
...
+
+* [Google Cloud Key Management](https://cloud.google.com/security-key-management?target=_blank).
+* [HashiCorp Vault](https://www.vaultproject.io/?target=_blank).
+* [KeyFactor](https://www.keyfactor.com/).
+
+For more information, check the help screen for the `gcp-kms-key`, `keyfactor-signserver-worker`, and `-vault-transit-key` options, available by running the command `mender-artifact sign --help`.
 
 
-## Supported cryptography standards
+## Hardware security
```
Comment (vpodzime) [inline 06.Artifact-creation/07.Sign-and-verify/docs.md:137]: :+1:

**Review by vpodzime — APPROVED** [review]
Looks good to me either way.

File: 06.Artifact-creation/07.Sign-and-verify/docs.md (hunk 4 above)
Comment (vpodzime) [reply 06.Artifact-creation/07.Sign-and-verify/docs.md:38]: I thought "supporting/having PKCS#11 interface" was equal to "compatible with `openssl`", but I don't know if that's 100% the case.

**Review by lluiscampos — APPROVED** [review]
Looks good to me. Thank you!

File: 02.Overview/11.Security/docs.md [hunk 6]
```diff
@@ -34,14 +34,20 @@ the Mender Client can connect to the server over HTTPS, you can schedule updates
...
 
-Each Device has a unique public and private RSA, ECDSA or ED25519 key. You can
-generate this offline and provision it with the Device storage, otherwise the
-Mender Client will automatically generate an RSA key pair with a default length
-of 3072 bits when it launches for the first time. Once generated, private key cannot be
-changed or retrieved by means of API calls. If you decide to re-generate the
-keys on the Device, it will require going through the authorization process
-again. The client passes the public key in authorization requests to the server and
-you can see the public key of a Device in the Mender UI.
+Each Device has a unique RSA or Elliptic Curve (P-256 or Ed25519) keypair (public
```
Comment (estenberg) [inline 02.Overview/11.Security/docs.md:37]: key pair

File: 02.Overview/11.Security/docs.md [hunk 7]
```diff
@@ -34,14 +34,20 @@ the Mender Client can connect to the server over HTTPS, you can schedule updates
...
+and private key). You can generate this offline and provision it with the Device
+storage, otherwise the Mender Client will automatically generate an Ed25519 key pair
+when it launches for the first time. Once generated, private key cannot be changed
+or retrieved by means of API calls. If you decide to re-generate the keys on the
+Device, it will require going through the authorization process again. The client
+passes the public key in authorization requests to the server and you can see the
+public key of a Device in the Mender UI.
+
+<!-- AUTOVERSION: "version %"/ignore -->
+!!!!! Mender Client version 4.1.0 and later generates an Ed25519 keypair by
```
Comment (estenberg) [inline 02.Overview/11.Security/docs.md:47]: key pair

File: 02.Overview/11.Security/docs.md [hunk 8]
```diff
@@ -34,14 +34,20 @@ the Mender Client can connect to the server over HTTPS, you can schedule updates
...
-changed or retrieved by means of API calls. If you decide to re-generate the
-keys on the Device, it will require going through the authorization process
-again. The client passes the public key in authorization requests to the server and
-you can see the public key of a Device in the Mender UI.
+Each Device has a unique RSA or Elliptic Curve (P-256 or Ed25519) keypair (public
+and private key). You can generate this offline and provision it with the Device
+storage, otherwise the Mender Client will automatically generate an Ed25519 key pair
+when it launches for the first time. Once generated, private key cannot be changed
+or retrieved by means of API calls. If you decide to re-generate the keys on the
+Device, it will require going through the authorization process again. The client
```
Comment (estenberg) [inline 02.Overview/11.Security/docs.md:42]: Client

File: 02.Overview/11.Security/docs.md [hunk 9]
```diff
@@ -34,14 +34,20 @@ the Mender Client can connect to the server over HTTPS, you can schedule updates
...
+passes the public key in authorization requests to the server and you can see the
+public key of a Device in the Mender UI.
+
+<!-- AUTOVERSION: "version %"/ignore -->
+!!!!! Mender Client version 4.1.0 and later generates an Ed25519 keypair by
+!!!!! default, while prior versions generated a 3072 bit RSA key by default.
```
Comment (estenberg) [inline 02.Overview/11.Security/docs.md:48]: key -> key pair

**Review by estenberg — CHANGES_REQUESTED** [review]
Just a few word changes. Feel free to merge after those.

**Review by estenberg — APPROVED** [review]

**Pull request merged**